WalletWasabi/Crypto/ZeroKnowledge/LinearRelation/Statement.cs

```diff
@@ -0,0 +1,46 @@
+using NBitcoin.Secp256k1;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.Crypto.ZeroKnowledge.LinearRelation
+{
+	public class Statement
+	{
+		public Statement(IEnumerable<Equation> equations)
+		{
+			// The equation matrix should not be jagged
+			Guard.NotNullOrEmpty(nameof(equations), equations);
+			var n = equations.First().Generators.Count();
+			Guard.True(nameof(equations), equations.All(e => e.Generators.Count() == n));
```

Comment:
maybe something like `CheckDimensions` can be done for non-jaggedness?

FWIW this could have been pushed down to the language by using `GroupElement[,]` instead of `IEnumerable<GroupElementVector`> but i think I prefer the advantages of the latter due to the readability of `Equation`. Maybe there is a middle ground?

Reply:
@lontivero mentioned in slack that maybe an `Equations` class would be appropriate.

Thinking about this a bit more, we could consider introducing `GroupElementMatrix` and `ScalarMatrix` as 2 dimensional arrays, and have `Equations` have a constructor that accepts a `GroupElementMatrix` where each row represents an equation, with the public input as the first component of every row vector and the generators as the remaining components.

I think this would simplify the construction and checking of these objects, and we can just have `Equations` also satisfy `IEnumerable<Equation>` as an interface, constructing these `Equation` objects on the fly using `new Equation(row.First(), row.Skip(1))`.

I'm not sure how I feel about this, and even if we agree that it's desirable then it should probably be a separate pull request.

Reply:
i don't think `GroupElementMatrix` is useful from an arithmetic point of view, in fact I would think that would only serve to obfuscate the current code if we do it as using matrix/vector multiplications (maybe it's just because I'm not entirely comfortable with that, I always have to re-derive the rules from first principles...). This might make sense later though if/when we some R1CS based proof system instead of Abe-Okhubo-Suzuki or Cramer-Damgard ring signatures for the rangeproofs.

The downside is like you said, yet another abstraction, which is why I'm ambivalent. The benefits would be:

1. more logical place for various guards
2. relying on on language level abstractions for dimension/rank checking (`GroupElement[,]` ensures it's non jagged)
3. possibly cleaner constructor syntax, but i haven't actually verified if this is legal C# syntax:
```cs
new Statement(
    new GroupElementMatrix {
        { P1, G11, G12, G13 ... },
        { P2, G21, G22, G23 ... },
        ...
   },
};
```

Reply:
FWIW if i'm not mistaken:

- the system of equations would be `Gw =? P` where `G` is a matrix of generators and `w` is a column vector using standard matrix multiplication
- public nonce generation would be `R = (k * G)1` where `*` is a Hadamard product (component wise) and 1 denotes a column vector of all 1's
- the response matrix `s = k + e*x` would be given by matrix addition of the secret nonce matrix k and e is just scalar product of the secret matrix x
- the verification equation would be `s*G =? R + eP`, again using a Hadamard product on the LHS

this would be a bit icky with overloads since we would need two product operators and one addition operator

Reply:
we agreed in the call that this is an overcomplication